appservice/src/main/java/nu/yona/server/subscriptions/rest/UserPhotoController.java

```diff
@@ -0,0 +1,151 @@
+	}
+
+	@RequestMapping(value = "/users/{userId}/photo", method = RequestMethod.PUT, consumes = "multipart/form-data")
+	@ResponseBody
+	public ResponseEntity<UserPhotoResource> uploadUserPhoto(
+			@RequestHeader(value = Constants.PASSWORD_HEADER) Optional<String> password,
+			@RequestParam(value = "file", required = false) MultipartFile userPhoto, @PathVariable UUID userId)
+	{
+		try (CryptoSession cryptoSession = CryptoSession.start(password, () -> userService.canAccessPrivateData(userId)))
+		{
+			return createUserPhotoUploadResponse(
+					userPhotoService.addUserPhoto(userId, UserPhotoDto.createUnsavedInstance(getPngBytes(userPhoto))));
+		}
+	}
+
+	@RequestMapping(value = "/users/{userId}/photo", method = RequestMethod.DELETE)
+	@ResponseBody
+	public ResponseEntity<Void> removeUserPhoto(@RequestHeader(value = Constants.PASSWORD_HEADER) Optional<String> password,
+			@PathVariable UUID userId)
+	{
+		try (CryptoSession cryptoSession = CryptoSession.start(password, () -> userService.canAccessPrivateData(userId)))
+		{
+			userPhotoService.removeUserPhoto(userId);
+			return new ResponseEntity<Void>(HttpStatus.OK);
+		}
+	}
+
+	private ResponseEntity<UserPhotoResource> createUserPhotoUploadResponse(UserPhotoDto userPhoto)
+	{
+		return new ResponseEntity<>(new UserPhotoResourceAssembler().toResource(userPhoto), HttpStatus.OK);
+	}
+
+	private static byte[] getPngBytes(MultipartFile file)
+	{
+		try
+		{
+			BufferedImage image = ImageIO.read(file.getInputStream());
+			if (image == null)
+			{
+				throw InvalidDataException.notSupportedPhotoFileType();
+			}
+			ByteArrayOutputStream pngBytes = new ByteArrayOutputStream();
+			ImageIO.write(image, "png", pngBytes);
+			return pngBytes.toByteArray();
+		}
+		catch (IOException e)
+		{
+			throw YonaException.unexpected(e);
+		}
+	}
+
+	public static ControllerLinkBuilder getUserPhotoLinkBuilder(UUID userPhotoId)
+	{
+		return linkTo(methodOn(UserPhotoController.class).getUserPhoto(userPhotoId));
+	}
+
+	public static class UserPhotoResource extends Resource<UserPhotoDto>
+	{
+		public UserPhotoResource(UserPhotoDto userPhoto)
+		{
+			super(userPhoto);
+		}
+	}
+
+	public static class UserPhotoResourceAssembler extends ResourceAssemblerSupport<UserPhotoDto, UserPhotoResource>
+	{
+		public UserPhotoResourceAssembler()
+		{
+			super(UserPhotoController.class, UserPhotoResource.class);
+		}
+
+		@Override
+		public UserPhotoResource toResource(UserPhotoDto userPhoto)
+		{
+			UserPhotoResource userPhotoResource = instantiateResource(userPhoto);
+			addSelfLink(userPhotoResource);
+			return userPhotoResource;
+		}
+
+		@Override
+		protected UserPhotoResource instantiateResource(UserPhotoDto userPhoto)
+		{
+			return new UserPhotoResource(userPhoto);
+		}
+
+		private void addSelfLink(Resource<UserPhotoDto> userPhotoResource)
+		{
+			userPhotoResource
+					.add(UserPhotoController.getUserPhotoLinkBuilder(userPhotoResource.getContent().getId()).withSelfRel());
```

Comment:
Strictly speaking, this is not a self-link. When you GET that link, you get the photo itself, not the ``UserPhoto`` entity. Can't we use the "yona:userPhoto" link relation instead, and omit the self-link?

Reply:
Will change